hunting/__main__.py

```diff
@@ -0,0 +1,278 @@
+def hunting():
+    """Commands for managing hunting queries and converting TOML to Markdown."""
+    pass
+
+
+@hunting.command('generate-markdown')
+@click.argument('path', required=False)
+def generate_markdown(path: click.Path):
+    """Convert TOML hunting queries to Markdown format."""
+    markdown_generator = MarkdownGenerator(HUNTING_DIR)
+
+    if path:
+        path = Path(path)
+        if path.is_file() and path.suffix == '.toml':
+            click.echo(f"Generating Markdown for single file: {path}")
+            markdown_generator.process_file(path)
+        elif (HUNTING_DIR / path).is_dir():
+            click.echo(f"Generating Markdown for folder: {path}")
+            markdown_generator.process_folder(path)
+        else:
+            click.echo(f"Invalid path: {path}. It should be a valid TOML file or a folder.")
+    else:
+        click.echo("Generating Markdown for all files.")
+        markdown_generator.process_all_files()
+
+    # After processing, update the index
+    markdown_generator.update_index_md()
+
+
+@hunting.command('refresh-index')
+def refresh_index():
+    """Refresh the index.yml file from TOML files and then refresh the index.md file."""
+    click.echo("Refreshing the index.yml and index.md files.")
+    update_index_yml(HUNTING_DIR)
+    markdown_generator = MarkdownGenerator(HUNTING_DIR)
+    markdown_generator.update_index_md()
+    click.echo("Index refresh complete.")
+
+
+@hunting.command('search')
+@click.option('--tactic', type=str, default=None, help="Search by MITRE tactic ID (e.g., TA0001)")
+@click.option('--technique', type=str, default=None, help="Search by MITRE technique ID (e.g., T1078)")
+@click.option('--sub-technique', type=str, default=None, help="Search by MITRE sub-technique ID (e.g., T1078.001)")
+@click.option('--data-source', type=str, default=None, help="Filter by data_source like 'aws', 'macos', or 'linux'")
+@click.option('--keyword', type=str, default=None, help="Search by keyword in name, description, and notes")
+def search_queries(tactic: str, technique: str, sub_technique: str, data_source: str, keyword: str):
+    """Search for queries based on MITRE tactic, technique, sub-technique, or data_source."""
+
+    if not any([tactic, technique, sub_technique, data_source, keyword]):
+        raise click.UsageError("""Please provide at least one filter (tactic, technique, sub-technique,
+                               data_source or keyword) to search queries.""")
+
+    click.echo("Searching for queries based on provided filters...")
+
+    # Create an instance of the QueryIndex class
+    query_index = QueryIndex(HUNTING_DIR)
+
+    # Filter out None values from the MITRE filter tuple
+    mitre_filters = tuple(filter(None, (tactic, technique, sub_technique)))
+
+    # Call the search method of QueryIndex with the provided MITRE filters, data_source, and keyword
+    results = query_index.search(mitre_filter=mitre_filters, data_source=data_source, keyword=keyword)
```

Comment:
btw, thank you very much for all the inline comments in the code, it is greatly appreciated!